src/rpc/util.cpp

```diff
@@ -634,6 +636,49 @@ UniValue RPCHelpMan::HandleRequest(const JSONRPCRequest& request) const
     return ret;
 }
 
+using CheckFn = void(const RPCArg&);
+static const UniValue* DetailMaybeArg(CheckFn* check, const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i)
+{
+    CHECK_NONFATAL(i < params.size());
+    const UniValue& arg{CHECK_NONFATAL(req)->params[i]};
+    const RPCArg& param{params.at(i)};
+    if (check) check(param);
+
+    if (!arg.isNull()) return &arg;
+    if (!std::holds_alternative<RPCArg::Default>(param.m_fallback)) return nullptr;
+    return &std::get<RPCArg::Default>(param.m_fallback);
+}
```

Comment:
With `MaybeArg` always working, could make this:

```c++
static const UniValue* DetailMaybeArg(const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i, bool null_ok)
 {
...
    if (!null_ok) CheckRequiredOrDefault(param);
...
}

#define TMPL_INST_MAYBE(ret_type, get_arg, return_code, return_code_null) \
    template <>                                                \
    ret_type RPCHelpMan::get_arg(size_t i) const               \
    {                                                          \
        const UniValue* maybe_arg{                             \
            DetailMaybeArg(m_args, m_req, i, true),            \
        };                                                     \
        return maybe_arg ? return_code : return_code_null;     \
    }                                                          \
    void force_semicolon()

#define TMPL_INST_REQ(ret_type, get_arg, return_code)          \
    template <>                                                \
    ret_type RPCHelpMan::get_arg(size_t i) const               \
    {                                                          \
        const UniValue* maybe_arg{                             \
            DetailMaybeArg(m_args, m_req, i, false),           \
        };                                                     \
        CHECK_NONFATAL(maybe_arg);                             \
        return return_code;                                    \
    }                                                          \
    void force_semicolon()
 
TMPL_INST_MAYBE(std::optional<double>, ArgValue<std::optional<double>>, std::optional{maybe_arg->get_real()}, std::nullopt);
TMPL_INST_REQ(int, ArgValue<int>, maybe_arg->getInt<int>());
```

Reply:
Okay, you nerdsniped me successfully. https://github.com/ajtowns/bitcoin/blob/9f332c93bcf7c2a1a4796dda4d2304014b10037e/src/rpc/util.cpp#L684-L705

Reply:
Yeah, pretty rare that a nerdsnipe results in anything productive. But you're not going to get me with C++20 features until #23363 is done...